src/Functional/Where.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace Functional;
+
+use Functional\Exceptions\InvalidArgumentException;
+
+/**
+ * Return the subset of objects which exactly match the supplied properties
+ *
+ * @param iterable|Traversable|array $objects
+ * @param array $propertyMap
+ * @return array
+ */
+function where($objects, array $propertyMap)
+{
+    /**
+     * Helper function
+     *
+     * Read a property of an object with fallbacks for arrays and methods
+     *
+     * @param $object
+     * @param $property
+     * @return mixed
+     */
+    $readProperty = function($object, $property) {
+        if (is_object($object)) {
+            if (method_exists($object, $property)) {
+                return $object->{$property}();
+            } elseif (property_exists($object, $property)) {
```

Comment:
property_exists() will return true for private properties. in_array() + get_class_vars() should work though.

Reply:
Isn't that a usefull error though "can't access private property"?

Reply:
Principle of the least surprise for me would mean for it to behave like a a non-existent property. Imagine you are using where() on a list and you catch the InvalidArgumentException because you foresee some property being missing. Now you go ahead and change the class that is used somewhere in you application and introduce a private property and suddenly you have a different error case. I would handle it consistently with the "non-existent property case".